### PR TITLE
Add a helper method for reading/writing memory

### PR DESCRIPTION
**Pull request by tenderlove**
This commit adds two new methods, `Fiddle::Pointer.read` and `Fiddle::Pointer.write`.  Both methods take an address, and will read or write bytes at that address respectively.

For example we can read from an address without making a Pointer object:

```ruby
Fiddle::Pointer.read(address, 5) # read 5 bytes
```

We can also write to an address without allocating a Pointer object:

```ruby
Fiddle::Pointer.write(address, "bytes") # write 5 bytes
```

This allows us to read / write memory at arbitrary addresses without instantiating a new `Fiddle::Pointer` object.

Examples where this API would be useful [1](https://github.com/tenderlove/tenderjit/blob/f03481d28bff4d248746e596929b0841de65f181/lib/tenderjit/fiddle_hacks.rb#L26-L28) [2](https://github.com/tenderlove/ruby/blob/77c8daa2d40dd58eeb3785ce17dea2ee38f308d1/lib/ruby_vm/rjit/c_pointer.rb#L193) [3](https://github.com/tenderlove/ruby/blob/77c8daa2d40dd58eeb3785ce17dea2ee38f308d1/lib/ruby_vm/rjit/c_pointer.rb#L284)

I also added a writer method for the same reasons as the reader.

### CODE REVIEW

**Review by kou — COMMENTED** [review]
Hmm. `Fiddle::Pointer[value]` and `Fiddle::Pointer[address, length]` may be confused.
How about `Fiddle::Pointer.read`/`Fiddle::Pointer.write` (like `IO.read`/`IO.write`) instead?

**Comment by tenderlove** [comment]
> Hmm. Fiddle::Pointer[value] and Fiddle::Pointer[address, length] may be confused.
How about Fiddle::Pointer.read/Fiddle::Pointer.write (like IO.read/IO.write) instead?

That sounds good to me.  I wish we could use `[]` and `[]=`, but I agree it's confusing.

File: ext/fiddle/pointer.c [hunk 1]
```diff
@@ -799,6 +799,33 @@ rb_fiddle_ptr_s_to_ptr(VALUE self, VALUE val)
...
+ *    Fiddle::Pointer.read(address, len)     => string
+ *
+ * Or read the memory at address +address+ with length +len+ and return a
+ * string with that memory
+ */
+
+static VALUE
+rb_fiddle_ptr_read_mem(VALUE klass, VALUE address, VALUE len)
+{
+    return rb_str_new((char *)NUM2ULONG(address), NUM2ULONG(len));
```
Comment (kou) [inline ext/fiddle/pointer.c:813]:
```suggestion
    return rb_str_new((char *)NUM2PTR(address), NUM2ULONG(len));
```

File: test/fiddle/test_pointer.rb [hunk 2]
```diff
@@ -10,6 +10,38 @@ def dlwrap arg
...
+      ptr = Fiddle::Pointer.new(address)
+      bytes_to_write = Fiddle::SIZEOF_VOIDP.times.to_a
+      bytes_to_write.each { |i| ptr[i] = i }
+
+      # Read the bytes out again
+      bytes = Fiddle::Pointer.read(address, Fiddle::SIZEOF_VOIDP).bytes
+      assert_equal bytes_to_write, bytes
+    ensure
+      Fiddle.free address
+    end
```
Comment (kou) [inline test/fiddle/test_pointer.rb:43]: How about unifying them? It seems that the first test covers the second one.

```suggestion
    def test_can_read_write_memory
      # Allocate some memory
      address = Fiddle.malloc(Fiddle::SIZEOF_VOIDP)

      bytes_to_write = Fiddle::SIZEOF_VOIDP.times.to_a.pack("C*")

      # Write to the memory
      Fiddle::Pointer.write(address, bytes_to_write)

      # Read the bytes out again
      bytes = Fiddle::Pointer.read(address, Fiddle::SIZEOF_VOIDP)
      assert_equal bytes_to_write, bytes
    ensure
      Fiddle.free address
    end
```

**Review by kou — APPROVED** [review]
+1

**Pull request merged**